### PR TITLE
Add script for generating occupancy maps for WorldForge worlds

### PR DESCRIPTION
**Pull request by mm318**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: scripts/genmap_script.sh [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+cd simulation_ws
+colcon build
+source install/local_setup.sh
+cd ..
+
+map_output_path=$(dirname $(dirname $world_source_path))/maps/map
+
+roslaunch cloudwatch_simulation start_map_service.launch &
+sleep 10
```
Comment (samuelgundry) [inline scripts/genmap_script.sh]: Instead of an arbitrary sleep, you can wait for the service to exist and/or timeout. This is more robust and avoids flakey race conditions.

File: simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager [hunk 2]
```diff
@@ -261,13 +261,14 @@ class RouteManager():
...
         rate = rospy.Rate(1)
         while not rospy.is_shutdown():
-            if self.bad_goal_counter > 10:
-                rospy.loginfo("Stopping route manager due to too many bad goals. Check that your occupancy map has 'Trinary' value representation and is not visually noisy/incorrect")
-                return
-            else:
+            try:
+                if self.bad_goal_counter > 10:
+                    raise ValueError("Bad goal limit exceeded")
+
```
Comment (samuelgundry) [inline simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager]: These are unrelated? Let's rebase on latest.

**Review by samuelgundry — CHANGES_REQUESTED** [review]
Post separate PRs for the unrelated changes

File: simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager (hunk 2 above)
Comment (mm318) [reply simulation_ws/src/aws_robomaker_simulation_common/nodes/route_manager]: Done.

File: scripts/genmap_script.sh (hunk 1 above)
Comment (mm318) [reply scripts/genmap_script.sh]: @ahtsan would you be able to address this issue?

**Comment by mm318** [comment]
@ahtsan and @ojasjoshi, can I please get a thorough review on the README changes to make sure that the section order and flow in the changes makes sense? Thanks!

**Comment by ojasjoshi** [comment]
> @ahtsan and @ojasjoshi, can I please get a thorough review on the README changes to make sure that the section order and flow in the changes makes sense? Thanks!

You mean the README changes in this PR?

File: README.md [hunk 3]
```diff
@@ -143,6 +143,14 @@ source simulation_ws/install/local_setup.sh
...
 ```
 
+By default, WorldForge packages will load the exported world. To override, specify the environment variable `WORLD_ID`. 
+
+```bash
+# use worldId found in "src/aws_robomaker_worldforge_worlds/worlds"
+# e.g, generation_05wq8sybdcn2_world_1
+export WORLD_ID=<worldId>  
+```
+
```
Comment (ojasjoshi) [inline README.md]: This whole section is not required (this is taken care of in the prerequisite step of map generation)

File: README.md [hunk 4]
```diff
@@ -193,46 +201,55 @@ and [create a simulation job](https://docs.aws.amazon.com/robomaker/latest/dg/cr
...
-```bash
-# Install dependencies (Ubuntu >18.04)
-sudo apt-get install ruby-dev libxml-xpath-perl libxml2-utils
-```
+Currently the following worlds are supported:
+- [`bookstore`](https://github.com/aws-robotics/aws-robomaker-bookstore-world)  
+- [`small_house`](https://github.com/aws-robotics/aws-robomaker-small-house-world)  
+- [`small_warehouse`](https://github.com/aws-robotics/aws-robomaker-small-warehouse-world)  
+- [`no_roof_small_warehouse`](https://github.com/aws-robotics/aws-robomaker-small-warehouse-world)
+- worldforge
```
Comment (ojasjoshi) [inline README.md]: this("worldforge") can be removed as the next section "Generate map for WorldForge worlds" explicitly gives instructions for worldforge

**Comment by mm318** [comment]
> You mean the README changes in this PR?

Right, the changes being made in this PR.

File: README.md [hunk 5]
```diff
@@ -193,46 +201,55 @@ and [create a simulation job](https://docs.aws.amazon.com/robomaker/latest/dg/cr
...
-rosdep install --from-paths src -r -y
-cd ..
+./scripts/genmap_script.sh <world_name>
 ```
+where `<world_name>` can be any value in the list above. Extra steps below are needed for `worldforge`.
+
+
+#### Generating map for WorldForge worlds
+
+After exporting a world from WorldForge, we can unzip the content and move under simulation_ws package:
```
Comment (ojasjoshi) [inline README.md]: Can be renamed to: After exporting a world from WorldForge, unzip and move the contents under simulation_ws workspace

File: README.md [hunk 6]
```diff
@@ -193,46 +201,55 @@ and [create a simulation job](https://docs.aws.amazon.com/robomaker/latest/dg/cr
 
 Procedurally generate an occupancy map for any gazebo world. This map can then be plugged in to navigate a robot in Worldforge worlds. For other aws-robotics worlds, this procedure is optional for the use-cases mentioned in this README. 
 
-### Pre-build
+### Script for default worlds/config
```
Comment (ojasjoshi) [inline README.md]: Can be renamed to: ##Generate map for a aws-robotics world with default config

File: README.md [hunk 7]
```diff
@@ -193,46 +201,55 @@ and [create a simulation job](https://docs.aws.amazon.com/robomaker/latest/dg/cr
...
-cd simulation_ws
-rosws update
-rosdep install --from-paths src -r -y
-cd ..
+./scripts/genmap_script.sh <world_name>
 ```
+where `<world_name>` can be any value in the list above. Extra steps below are needed for `worldforge`.
+
+
+#### Generating map for WorldForge worlds
```
Comment (ojasjoshi) [inline README.md]: Can be renamed to: ##Generate map for a Worldforge world with default config

File: README.md [hunk 8]
```diff
@@ -193,46 +201,55 @@ and [create a simulation job](https://docs.aws.amazon.com/robomaker/latest/dg/cr
...
+
+#For worldforge worlds, set WORLD_ID to the name of your WF exported world (eg: generation_40r67s111n9x_world_3),
+export WORLD_ID=<worldforge-world-name>
+
+# Run map generation script
+./scripts/genmap_script.sh worldforge
 ```
 
-### Generate Occupancy Map via map generation plugin
+### Custom worlds/config
```
Comment (ojasjoshi) [inline README.md]: Can be renamed to: ##Generate map for a custom world with custom config

File: README.md [hunk 9]
```diff
@@ -193,46 +201,55 @@ and [create a simulation job](https://docs.aws.amazon.com/robomaker/latest/dg/cr
...
 Procedurally generate an occupancy map for any gazebo world. This map can then be plugged in to navigate a robot in Worldforge worlds. For other aws-robotics worlds, this procedure is optional for the use-cases mentioned in this README. 
 
-### Pre-build
+### Script for default worlds/config
 
-```bash
-# Install dependencies (Ubuntu >18.04)
-sudo apt-get install ruby-dev libxml-xpath-perl libxml2-utils
-```
+Currently the following worlds are supported:
```
Comment (ojasjoshi) [inline README.md]: Can be renamed to: Currently, the following aws-robotics worlds are supported,

File: README.md (hunk 3 above)
Comment (ahtsan) [reply README.md]: This part belongs to `Running simulation with a WorldForge world`. However, since map generation now is NOT optional, the prerequisite step of map generation will be merged in here so this part could be removed.

File: README.md (hunk 4 above)
Comment (ahtsan) [reply README.md]: right, `WorldForge` is not part of `aws-robotics` worlds.

File: scripts/genmap_script.sh (hunk 1 above)
Comment (ahtsan) [reply scripts/genmap_script.sh]: working on it, seem tricky to do it in a shell script

Comment (ahtsan) [reply scripts/genmap_script.sh]: I added this issue by using a python snippet with `rospy.wait_for_service()`

File: README.md (hunk 7 above)
Comment (ahtsan) [reply README.md]: Fixed

File: README.md (hunk 6 above)
Comment (ahtsan) [reply README.md]: Fixed

File: README.md (hunk 5 above)
Comment (ahtsan) [reply README.md]: Fixed

File: README.md (hunk 8 above)
Comment (ahtsan) [reply README.md]: Fixed

File: README.md (hunk 9 above)
Comment (ahtsan) [reply README.md]: Fixed

**Comment by ahtsan** [comment]
@ojasjoshi Comment addressed for README, please take a look again.

**Comment by ojasjoshi** [comment]
README changes look good to me. 
Just wanted to check once if removing turtlebot3_description from the dependency is still intended and in-line with our latest knowledge from hello-world SA changes.

**Comment by ahtsan** [comment]
> README changes look good to me.
> Just wanted to check once if removing turtlebot3_description from the dependency is still intended and in-line with our latest knowledge from hello-world SA changes.

Good point. Yes this is intended, We should use turtlebot3_navigation instead of turtlebot3_description in CloudWatch SA because of the navigation stack.

**Comment by ahtsan** [comment]
@ojasjoshi @mm318 @samuelgundry Updated README and script to wait for service to be available, instead of a ad-hoc 10 second wait. Please take a look.

**Comment by ojasjoshi** [comment]
Looks good to me

**Comment by ojasjoshi** [comment]
Looks good to me.. only one suggestion in the README

File: README.md [hunk 10]
```diff
@@ -118,11 +118,13 @@ Launch the application with the following commands:
...
+For navigation, you can generate a map with map generation plugin. See [this](#generate-occupancy-map-via-map-generation-plugin) for instructions.
+
 ![CloudWatchMetrics01.png](docs/images/BookstoreRVizPlan01.png)
 
 ### Run with a AWS Robomaker WorldForge world
 
-Pre-requisite: Generate a map for your worldforge exported world following these [instructions](#generate-map-for-your-world).
+Pre-requisite: Generate a map for your worldforge exported world following these [instructions](#generate-map-for-a-worldforge-world-with-default-config).
 
 Move the generated map file to cloudwatch_simulation package,
```
Comment (ojasjoshi) [inline README.md]: This(line 129) might be redundant and/or confusing if the genmap_script is used for map generation.. maybe could add a note clarifying this

Comment (ahtsan) [reply README.md]: Good point, we should probably remove this part as this is covered in `genmap_script.sh`

Comment (ahtsan) [reply README.md]: We should move this part under `Generate map for a custom world with custom config` as it's a required step

Comment (ojasjoshi) [reply README.md]: Yep looks good now

File: map_config/default_args.json [hunk 11]
```diff
@@ -1 +1,22 @@
...
+        "./map_config/small_warehouse.rb",
+        "aws-robomaker-small-warehouse-world"
+    ],
+    "small_house": [
+        "./map_config/small_house.rb",
+        "aws-robomaker-small-house-world"
+    ],
+    "worldforge": [
+        "./map_config/worldforge.rb",
+        "./simulation_ws/src/aws_robomaker_worldforge_pkgs/aws_robomaker_worldforge_worlds/worlds"
```
Comment (samuelgundry) [inline map_config/default_args.json:20]: Why does this config have `simulation_ws/src` prefix and the others don't?

File: simulation_ws/src/cloudwatch_simulation/CMakeLists.txt [hunk 12]
```diff
@@ -3,17 +3,17 @@ project(cloudwatch_simulation)
...
   DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}
 )
 
-# Copy the rviz model for easier access in AWS RoboMaker RViz 
-install(FILES ${turtlebot3_description_DIR}/../rviz/model.rviz
+# Copy the rviz model for easier access in RoboMaker RViz 
+install(FILES ${turtlebot3_navigation_DIR}/../rviz/turtlebot3_navigation.rviz
   DESTINATION ${CATKIN_PACKAGE_SHARE_DESTINATION}/rviz
-  RENAME turtlebot3_model.rviz
+  RENAME turtlebot3_navigation.rviz
```
Comment (samuelgundry) [inline simulation_ws/src/cloudwatch_simulation/CMakeLists.txt]: FWIW, `turtlebot3_navigation.rviz` is *not* the same as `turtlebot3_model.rviz`.  They configure rviz differently.  I don't think it's a big deal -- we want easier/better access from rviz tool to configs on disk/in the bundle sooner than later, which would make this `install()` work-around less important.

File: simulation_ws/src/cloudwatch_simulation/package.xml [hunk 13]
```diff
@@ -17,8 +17,8 @@
   <depend>tf</depend>
   <depend>gazebo_ros</depend>
   <depend>gazebo_plugins</depend>
-  <!-- required to install .rviz model-->
-  <depend>turtlebot3_description</depend>
```
Comment (samuelgundry) [inline simulation_ws/src/cloudwatch_simulation/package.xml:21]: This is not the same -- turtlebot3_description is still required, yes?  It provides the URDF, the robot 3D model.  

Separately, why does *simulation* require turtlebot3_navigation stack?  Maybe this is a legacy bug?

Comment (ahtsan) [reply simulation_ws/src/cloudwatch_simulation/package.xml:21]: We can keep `turtlebot3_model.rviz` instead of `turtlebot3_navigation.rviz` to avoid having `turtlebot3_navigation` as a dependency in simulation workspace

Comment (ahtsan) [reply simulation_ws/src/cloudwatch_simulation/package.xml:21]: I thought we always want to install `turtlebot3_navigation.rviz` when navigation is available, e.g. 
https://github.com/aws-robotics/aws-robomaker-bookstore-world/blob/ros1/CMakeLists.txt#L16
https://github.com/aws-robotics/aws-robomaker-small-house-world/blob/ros1/CMakeLists.txt#L17

File: map_config/default_args.json (hunk 11 above)
Comment (ahtsan) [reply map_config/default_args.json:20]: the python script that loads the .rb file handles aws-robotics worlds and WorldForge world differently, see
https://github.com/aws-robotics/aws-robomaker-sample-application-cloudwatch/blob/anson/worldforge-navigation/scripts/add_map_plugin.py#L36

File: simulation_ws/src/cloudwatch_simulation/package.xml (hunk 13 above)
Comment (ahtsan) [reply simulation_ws/src/cloudwatch_simulation/package.xml:21]: We can keep `turtlebot3_description` at this moment to unblock this PR.

**Comment by ahtsan** [comment]
@samuelgundry Comment addressed, please take another look

**Review by samuelgundry — APPROVED** [review]
Thanks for decoupling the tb3 navigation vs. description.rviz change.
